### PR TITLE
Avoid creating distributions when inferring value domain

### PR DESCRIPTION
**Pull request by fritzo**
Addresses #412
pair coded with @eb8680 

This refactors logic in `Distribution._infer_value_domain()` to avoid creating a temporary dummy distribution *if* backend distributions implement a `.infer_shapes()` method.

### CODE REVIEW

File: funsor/torch/distributions.py [hunk 1]
```diff
@@ -115,16 +115,24 @@ def _infer_value_domain(**kwargs):
     return kwargs['v']
 
 
+@methodof(Categorical)  # noqa: F821
```
Comment (eb8680) [inline funsor/torch/distributions.py:118]: Add `methodof(CategoricalLogits)`

**Comment by fritzo** [comment]
Test failures appear unrelated.
```
...
from jaxlib import cusolver[0m
ImportError: cannot import name 'cusolver'[0m
Makefile:46: recipe for target 'test' failed
```
@fehiepsi any suggestions?

**Comment by eb8680** [comment]
Looks like a new JAX bug? https://github.com/google/jax/issues/5374

**Comment by eb8680** [comment]
@fehiepsi it looks like a different new JAX error related to hashing of `DeviceArray`s is causing the build to fail, any idea what's going on?

**Comment by fehiepsi** [comment]
@fritzo Could you allow me rerun the build? Sometimes I would like to do it, for example to return this PR with the new jax release.

**Comment by fehiepsi** [comment]
@eb8680 Opps, I missed your comment. Will talk a look soon...

**Comment by fritzo** [comment]
> @fritzo Could you allow me to rerun the build?

@fehiepsi sure, feel free to rerun the builds. Do I need to authorize you on travis?

File: funsor/memoize.py [hunk 2]
```diff
@@ -17,7 +17,7 @@ def memoize(cache=None):
 
     @interpreter.interpretation(interpreter._INTERPRETATION)  # use base
     def memoize_interpretation(cls, *args):
-        key = (cls,) + tuple(id(arg) if (type(arg).__name__ == "DeviceArray") or not isinstance(arg, Hashable)
+        key = (cls,) + tuple(id(arg) if ("DeviceArray" in type(arg).__name__) or not isinstance(arg, Hashable)
```
Comment (fehiepsi) [inline funsor/memoize.py:20]: At some point ago, `type(jax.numpy.ones(3)).__name__` is `"_DeviceArray"`

File: funsor/tensor.py [hunk 3]
```diff
@@ -88,6 +88,16 @@ def __call__(cls, data, inputs=None, dtype="real"):
...
             data = data.__array__()
+        # XXX: in jaxlib 0.1.58 + jax 0.2.8, sometime opt_einsum returns
+        # a jaxlib.xla_client.Buffer, which is a type not supported in Tensor.
+        # It is unclear whether this is an issue. The good thing is: under jit,
+        # data will be a ShapedArray, so we won't go to this branch to do
+        # the extra job `jax.numpy.array(data)`.
+        elif "Buffer" in type(data).__name__ and get_backend() == "jax":
+            import jax
+
+            data = jax.numpy.array(data)
```
Comment (eb8680) [inline funsor/tensor.py:99]: Do we need to add something like this to `funsor.jax.ops.einsum` as well?

Comment (fehiepsi) [reply funsor/tensor.py:99]: @hawkinsp In https://github.com/google/jax/pull/2839, you mentioned that `xla_client.Buffer` is going to be deleted. But in the latest jax+jaxlib release, we are seeing that `Buffer` starts appearing (it didn't appear in previous releases). I just wonder if it is will be supported in the long run?

Comment (hawkinsp) [reply funsor/tensor.py]: Ah, this is a different `Buffer`. The one you are talking about was a Python class. The new one is a C++ class.

To improve dispatch times, we are moving a few parts of the JAX dispatch logic into C++, and in some cases (notably, uncomplicated `jit` calls), you will get a pure C++ object out (`Buffer`) without the additional Python wrapper class. It should duck type the same (e.g., have the same methods and act the same in most respects).

I believe both the Python and C++ variants report themselves as being instances of `DeviceArray`. Eventually we will only have the C++ version.

**Comment by fritzo** [comment]
@fehiepsi thanks for the help!

File: funsor/tensor.py (hunk 3 above)
Comment (fehiepsi) [reply funsor/tensor.py]: Thanks for the very clear explanation, @hawkinsp! I understand what's going on now.

**Pull request merged**